redoing PR Api objectlinks #33738

JonBendtsen opened the pull request:
# NEW [*API for links between objects*]
This PR introduces a new API endpoint which can GET, DELETE and POST an link between other objects inside Dolibarr.

Commit:
redoing PR Api objectlinks #33738

Commit:
loading objectlinks api from main api folder

JonBendtsen commented:
lots of pictures, comments and some json can be seen here

https://github.com/Dolibarr/dolibarr/pull/33738

mdeweerd commented on htdocs/api/class/api_objectlinks.class.php:
```diff
@@ -0,0 +1,473 @@
+class ObjectLinks extends DolibarrApi
+{
+	/**
+	 * @var array   $FIELDS     Mandatory fields, checked when create and update object
```
Using on "older" template I think - the variable name is not required in the property documentation.
Also, specify fields as `string[]` for static analysis

JonBendtsen replied:
@mdeweerd "older" ... I originally used v21 as my development base

mdeweerd replied:
v21 is "older" than develop .... 😄, 2546 commits behind develop.
I can't find a single `@var array   $FIELDS` in develop .

mdeweerd commented on htdocs/api/class/api_objectlinks.class.php:
```diff
@@ -0,0 +1,473 @@
+	 * Delete an object link
+	 *
+	 * @param   int     $id         object link ID
+	 * @return  array
```
Check some of the other api files to see examples of more specific type hinting for phpstan and phan for return values.
You can also find example for the type hinting of other arguments (request_data, etc) for other methods.

mdeweerd replied:
You'll find some examples in htdocs/adherents/class/api_members.class.php  .

JonBendtsen replied:
@mdeweerd  Thanks. I am not sure how I can fix these or even where the actual source of the issue is?

```
Error: Property ObjectLink::$fk_source (int) does not accept array<string, mixed>|string.
Error: Property ObjectLink::$fk_target (int) does not accept array<string, mixed>|string.
Error: Property ObjectLink::$fk_source (int) does not accept array<string, mixed>|string.
Error: Property ObjectLink::$fk_target (int) does not accept array<string, mixed>|string.
Error: Process completed with exit code 1.
```

JonBendtsen replied:
I have a suspicion that it could be the _checkvalforapi call, but I'd like some guidance

JonBendtsen replied:
no, it can't be, since it is called 3 times, and I only see the fk_source and target twice.

mdeweerd replied:
`_checkValForAPI` is mainly using `sanitizeVal`which is returning strings.
So we should not update the typehint for `_checkValForAPI`.
If I had full control, I would update `_checkValForAPI` to return 'int' when the 
`$fields` specification indicates it is an int.

Therefore, to respect the typing, we need to cast the values locally.

```php
$intFieldNames = array('fk_source','fk_target');
// [...]
$cleanVal = _checkValForAPI(...);
if (in_array($field, $intFieldNames)) {
  $this->objectlink->$field = (int) $cleanVal;
} else {
  $this->objectLink->$field = $cleanVal;
}
```

There is the same code issue in create, but there the static tools can not detect it because they do not know the potential values for $fields.

In getByValues and in deleteByValues you define an array with fk_source and fk_target, so the static check sees these key values and makes specific checks only for those.

This is why you get these notices two times for fk_source and fk_target only.


Idealy, the values would be cleaned and filtered in $this_validate, and the result used to fill the objectlinks values.

And to help static analysis, the return value of _validate could contain the exact typing of the return value example: `array{fk_source?:int,fk_target?:int,other:string,another:int}` .

In some API classes there are sevaral _validateXXX() functions, one per type.  That is also a way to ensure that you can have declare a more specific return type.

In doing so, you enable the tools to help you (and other developers) to ensure that the typings are optimum at all times.

JonBendtsen replied:
@mdeweerd  thank you, it builds cleanly now :-)

JonBendtsen replied:
@eldy  I think this is finally ready :-)

JonBendtsen replied:
@mdeweerd 

Maybe not as ready as I though - when just visiting the API explorer or reloading the page I get 

`500 : http://localhost:8080/api/index.php/explorer/swagger.json`

Apache logs from the container
```
JonSweet16:dolibarr_devtainer jonbendtsen$ podman logs -f --tail 0 dolibarr_develop | tee /tmp/api.load
[Sat Apr 12 19:43:21.260380 2025] [php:error] [pid 68:tid 68] [client 192.168.127.1:37977] PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 20480 bytes) in /var/www/html/includes/restler/framework/Luracast/Restler/CommentParser.php on line 208, referer: http://localhost:8080/api/index.php/explorer/
192.168.127.1 - - [12/Apr/2025:19:43:19 +0200] "GET /api/index.php/explorer/swagger.json HTTP/1.1" 500 363 "http://localhost:8080/api/index.php/explorer/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/134.0.0.0 Safari/537.36"
```

So I doubled the PHP memory to 512M, and it gave the same error message
```
[Sat Apr 12 19:48:31.714068 2025] [php:error] [pid 45:tid 45] [client 192.168.127.1:27619] PHP Fatal error:  Allowed memory size of 536870912 bytes exhausted (tried to allocate 20480 bytes) in /var/www/html/includes/restler/framework/Luracast/Restler/CommentParser.php on line 511, referer: http://localhost:8080/api/index.php/explorer/
192.168.127.1 - - [12/Apr/2025:19:48:29 +0200] "GET /api/index.php/explorer/swagger.json HTTP/1.1" 500 363 "http://localhost:8080/api/index.php/explorer/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/134.0.0.0 Safari/537.36"
```


and Dolibarr.log

```
2025-04-12 17:43:19 NOTICE  192.168.127.1        68     33 --- Access to GET /api/index.php/explorer/swagger.json - action=, massaction= NOTOKENRENEWAL=1
2025-04-12 17:43:19 DEBUG   192.168.127.1        68     33 sql=SELECT transkey, transvalue FROM llx_overwrite_trans where (lang='en_DK' OR lang IS NULL) AND entity IN (0, 0,1) ORDER BY lang DESC
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Scan directory /var/www/html/core/modules/ for module descriptor files, then search for API files
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modMrp.class.php - module=mrp - modulenameforenabled=mrp - moduledirforclass=mrp
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modWorkflow.class.php - module=workflow - modulenameforenabled=workflow - moduledirforclass=workflow
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modECM.class.php - module=ecm - modulenameforenabled=ecm - moduledirforclass=ecm
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modCommande.class.php - module=commande - modulenameforenabled=commande - moduledirforclass=commande
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modSalaries.class.php - module=salaries - modulenameforenabled=salaries - moduledirforclass=salaries
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modFTP.class.php - module=ftp - modulenameforenabled=ftp - moduledirforclass=ftp
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modEventOrganization.class.php - module=eventorganization - modulenameforenabled=eventorganization - moduledirforclass=eventorganization
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modNotification.class.php - module=notification - modulenameforenabled=notification - moduledirforclass=notification
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modPrelevement.class.php - module=prelevement - modulenameforenabled=prelevement - moduledirforclass=prelevement
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modTicket.class.php - module=ticket - modulenameforenabled=ticket - moduledirforclass=ticket
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modPaypal.class.php - module=paypal - modulenameforenabled=paypal - moduledirforclass=paypal
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modAgenda.class.php - module=agenda - modulenameforenabled=agenda - moduledirforclass=comm/action
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modBookmark.class.php - module=bookmark - modulenameforenabled=bookmark - moduledirforclass=bookmark
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modBookCal.class.php - module=bookcal - modulenameforenabled=bookcal - moduledirforclass=bookcal
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modKnowledgeManagement.class.php - module=knowledgemanagement - modulenameforenabled=knowledgemanagement - moduledirforclass=knowledgemanagement
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modService.class.php - module=service - modulenameforenabled=service - moduledirforclass=service
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modDav.class.php - module=dav - modulenameforenabled=dav - moduledirforclass=dav
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modWebsite.class.php - module=website - modulenameforenabled=website - moduledirforclass=website
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modLoan.class.php - module=loan - modulenameforenabled=loan - moduledirforclass=loan
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modBanque.class.php - module=banque - modulenameforenabled=banque - moduledirforclass=compta/bank
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modAccounting.class.php - module=accounting - modulenameforenabled=accounting - moduledirforclass=accountancy
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modDocumentGeneration.class.php - module=documentgeneration - modulenameforenabled=documentgeneration - moduledirforclass=documentgeneration
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modExpedition.class.php - module=expedition - modulenameforenabled=expedition - moduledirforclass=expedition
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modIntracommreport.class.php - module=intracommreport - modulenameforenabled=intracommreport - moduledirforclass=intracommreport
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modDataPolicy.class.php - module=datapolicy - modulenameforenabled=datapolicy - moduledirforclass=datapolicy
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modApi.class.php - module=api - modulenameforenabled=api - moduledirforclass=api
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modPaybox.class.php - module=paybox - modulenameforenabled=paybox - moduledirforclass=paybox
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modZapier.class.php - module=zapier - modulenameforenabled=zapier - moduledirforclass=zapier
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modProduct.class.php - module=product - modulenameforenabled=product - moduledirforclass=product
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modCollab.class.php - module=collab - modulenameforenabled=collab - moduledirforclass=collab
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modBom.class.php - module=bom - modulenameforenabled=bom - moduledirforclass=bom
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modUser.class.php - module=user - modulenameforenabled=user - moduledirforclass=user
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modPropale.class.php - module=propale - modulenameforenabled=propal - moduledirforclass=comm/propal
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modWebhook.class.php - module=webhook - modulenameforenabled=webhook - moduledirforclass=webhook
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modWorkstation.class.php - module=workstation - modulenameforenabled=workstation - moduledirforclass=workstation
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modPartnership.class.php - module=partnership - modulenameforenabled=partnership - moduledirforclass=partnership
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modGeoIPMaxmind.class.php - module=geoipmaxmind - modulenameforenabled=geoipmaxmind - moduledirforclass=geoipmaxmind
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modContrat.class.php - module=contrat - modulenameforenabled=contrat - moduledirforclass=contrat
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modPaymentByBankTransfer.class.php - module=paymentbybanktransfer - modulenameforenabled=paymentbybanktransfer - moduledirforclass=paymentbybanktransfer
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modAsset.class.php - module=asset - modulenameforenabled=asset - moduledirforclass=asset
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modRecruitment.class.php - module=recruitment - modulenameforenabled=recruitment - moduledirforclass=recruitment
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modFicheinter.class.php - module=ficheinter - modulenameforenabled=intervention - moduledirforclass=fichinter
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modSupplierProposal.class.php - module=supplierproposal - modulenameforenabled=supplier_proposal - moduledirforclass=supplier_proposal
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modModuleBuilder.class.php - module=modulebuilder - modulenameforenabled=modulebuilder - moduledirforclass=modulebuilder
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modEmailCollector.class.php - module=emailcollector - modulenameforenabled=emailcollector - moduledirforclass=emailcollector
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modFournisseur.class.php - module=fournisseur - modulenameforenabled=fournisseur - moduledirforclass=fourn
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modOpenSurvey.class.php - module=opensurvey - modulenameforenabled=opensurvey - moduledirforclass=opensurvey
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modTakePos.class.php - module=takepos - modulenameforenabled=takepos - moduledirforclass=takepos
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modMargin.class.php - module=margin - modulenameforenabled=margin - moduledirforclass=margin
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modStock.class.php - module=stock - modulenameforenabled=stock - moduledirforclass=product/stock
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modAdherent.class.php - module=adherent - modulenameforenabled=adherent - moduledirforclass=adherents
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modResource.class.php - module=resource - modulenameforenabled=resource - moduledirforclass=resource
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modCron.class.php - module=cron - modulenameforenabled=cron - moduledirforclass=cron
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modExternalSite.class.php - module=externalsite - modulenameforenabled=externalsite - moduledirforclass=externalsite
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modMultiCurrency.class.php - module=multicurrency - modulenameforenabled=multicurrency - moduledirforclass=multicurrency
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modImport.class.php - module=import - modulenameforenabled=import - moduledirforclass=import
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modFacture.class.php - module=facture - modulenameforenabled=facture - moduledirforclass=compta/facture
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modOpenIDConnect.class.php - module=openidconnect - modulenameforenabled=openidconnect - moduledirforclass=openidconnect
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modCategorie.class.php - module=categorie - modulenameforenabled=categorie - moduledirforclass=categories
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modMailmanSpip.class.php - module=mailmanspip - modulenameforenabled=mailmanspip - moduledirforclass=mailmanspip
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modIncoterm.class.php - module=incoterm - modulenameforenabled=incoterm - moduledirforclass=incoterm
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modGravatar.class.php - module=gravatar - modulenameforenabled=gravatar - moduledirforclass=gravatar
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modReception.class.php - module=reception - modulenameforenabled=reception - moduledirforclass=reception
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modTax.class.php - module=tax - modulenameforenabled=tax - moduledirforclass=tax
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modLabel.class.php - module=label - modulenameforenabled=label - moduledirforclass=label
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modVariants.class.php - module=variants - modulenameforenabled=variants - moduledirforclass=variants
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modStripe.class.php - module=stripe - modulenameforenabled=stripe - moduledirforclass=stripe
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modBarcode.class.php - module=barcode - modulenameforenabled=barcode - moduledirforclass=barcode
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modDon.class.php - module=don - modulenameforenabled=don - moduledirforclass=don
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modLdap.class.php - module=ldap - modulenameforenabled=ldap - moduledirforclass=ldap
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modExternalRss.class.php - module=externalrss - modulenameforenabled=externalrss - moduledirforclass=externalrss
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modClickToDial.class.php - module=clicktodial - modulenameforenabled=clicktodial - moduledirforclass=clicktodial
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modFckeditor.class.php - module=fckeditor - modulenameforenabled=fckeditor - moduledirforclass=fckeditor
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modSocialNetworks.class.php - module=socialnetworks - modulenameforenabled=socialnetworks - moduledirforclass=socialnetworks
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modHRM.class.php - module=hrm - modulenameforenabled=hrm - moduledirforclass=hrm
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modProjet.class.php - module=projet - modulenameforenabled=projet - moduledirforclass=projet
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modOauth.class.php - module=oauth - modulenameforenabled=oauth - moduledirforclass=oauth
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modSyslog.class.php - module=syslog - modulenameforenabled=syslog - moduledirforclass=syslog
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modComptabilite.class.php - module=comptabilite - modulenameforenabled=comptabilite - moduledirforclass=comptabilite
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modWebServices.class.php - module=webservices - modulenameforenabled=webservices - moduledirforclass=webservices
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modProductBatch.class.php - module=productbatch - modulenameforenabled=productbatch - moduledirforclass=productbatch
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modStockTransfer.class.php - module=stocktransfer - modulenameforenabled=stocktransfer - moduledirforclass=stocktransfer
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modExport.class.php - module=export - modulenameforenabled=export - moduledirforclass=export
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modPrinting.class.php - module=printing - modulenameforenabled=printing - moduledirforclass=printing
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modDeplacement.class.php - module=deplacement - modulenameforenabled=deplacement - moduledirforclass=deplacement
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modBlockedLog.class.php - module=blockedlog - modulenameforenabled=blockedlog - moduledirforclass=blockedlog
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modHoliday.class.php - module=holiday - modulenameforenabled=holiday - moduledirforclass=holiday
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modWebPortal.class.php - module=webportal - modulenameforenabled=webportal - moduledirforclass=webportal
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modAi.class.php - module=ai - modulenameforenabled=ai - moduledirforclass=ai
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modReceiptPrinter.class.php - module=receiptprinter - modulenameforenabled=receiptprinter - moduledirforclass=receiptprinter
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modDynamicPrices.class.php - module=dynamicprices - modulenameforenabled=dynamicprices - moduledirforclass=dynamicprices
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modSociete.class.php - module=societe - modulenameforenabled=societe - moduledirforclass=societe
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modDebugBar.class.php - module=debugbar - modulenameforenabled=debugbar - moduledirforclass=debugbar
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modExpenseReport.class.php - module=expensereport - modulenameforenabled=expensereport - moduledirforclass=expensereport
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Found module file modMailing.class.php - module=mailing - modulenameforenabled=mailing - moduledirforclass=mailing
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Scan directory /var/www/html/recruitment/core/modules/ for module descriptor files, then search for API files
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Scan directory /var/www/html/knowledgemanagement/core/modules/ for module descriptor files, then search for API files
2025-04-12 17:43:19 INFO    192.168.127.1        68     33 Scan directory /var/www/html/partnership/core/modules/ for module descriptor files, then search for API files
2025-04-12 17:43:21 INFO    192.168.127.1        68     33 --- End access to /api/index.php/explorer/swagger.json
 ```

mdeweerd replied:
Yes, repeated here, from CII and result visualized in browser.
So related to phpdoc.

![{CF944F4A-70E5-4DF8-BC9D-C3BC7828F501}](https://github.com/user-attachments/assets/1f40302f-df9a-40ac-91a2-fe915152126b)

mdeweerd replied:
I found the reason, suspected it : the types that are used for the param and return fields are used by Luracast/Restler and should not be "too complex".
However I learned on https://restler4.luracast.com/param.html that you can add some constraints with other tags. 

Made a PR to your fork.

JonBendtsen replied:
@mdeweerd  thanks, no errors when doing API explorer now :-) I'll look at phpstan because it now says:

```
Error: Property ObjectLink::$relationtype (string) does not accept int.
Error: Property ObjectLink::$sourcetype (string) does not accept int.
Error: Property ObjectLink::$targettype (string) does not accept int.
Error: Property ObjectLink::$relationtype (string) does not accept int.
Error: Property ObjectLink::$sourcetype (string) does not accept int.
Error: Property ObjectLink::$targettype (string) does not accept int.
```

JonBendtsen replied:
It is shown twice, and the relationtype is not used many places, so mostly looking at lines where it is assigned.

`private function _validate($data)` does it, but it is called 3 times.

This one is also done 3 times

```
				if (in_array($field, ObjectLinks::$INTFIELDS)) {
					$this->objectlink->$field = (int) $clean_field;
				} else {
					$this->objectlink->$field = (string) $clean_field;
				}
```

I wish phpstan would write a line number

mdeweerd replied:
phpstan and phan provide line numbers, but github is hiding them in the logs.

You can have a look at the PR's action summary where 10 notices of errors/warnings of the tools are shown with their line numbers and you can click to go to the location (which works fine if the file is expanded by default).

https://github.com/Dolibarr/dolibarr/actions/runs/14423020321

You can also download the artifacts from the analysis jobs which contain either the xml with the notices (and the line numbers) or text log.

JonBendtsen replied:
@mdeweerd  thank you, I see the line errors in the download.

JonBendtsen replied:
@mdeweerd I don't understand why phpstan complains with what it does where it does it :-(

it complains about line 250 and line 319 - but they are the same

https://github.com/JonBendtsen/dolibarr/blob/from_develop_api_objectlinks/htdocs/api/class/api_objectlinks.class.php#L250

The errors
```
  <error line="250" column="1" severity="error" message="Property ObjectLink::$relationtype (string) does not accept int." source="assign.propertyType" />
  <error line="250" column="1" severity="error" message="Property ObjectLink::$sourcetype (string) does not accept int." source="assign.propertyType" />
  <error line="250" column="1" severity="error" message="Property ObjectLink::$targettype (string) does not accept int." source="assign.propertyType" />
  ```
  
  but none of those are in the array that I check for, so they should be assigned to the else part of the if??
  
  ```
  
		foreach ($request_data as $field => $value) {
			$clean_field = $this->_checkValForAPI($field, $value, $this->objectlink);
			if (in_array($field, ObjectLinks::$INTFIELDS)) {
				$this->objectlink->$field = (int) $clean_field;
			} else {
				$this->objectlink->$field = (string) $clean_field;
			}
		}
```

JonBendtsen replied:
I changed the code to

```
		foreach ($request_data as $field => $value) {
			$clean_field = $this->_checkValForAPI($field, $value, $this->objectlink);
			if (in_array($field, ObjectLinks::$INTFIELDS)) {
				$this->objectlink->$field = (int) $clean_field;
				dol_syslog("integer and ".$field, LOG_DEBUG);
			} else {
				dol_syslog("string and ".$field, LOG_DEBUG);
				$this->objectlink->$field = (string) $clean_field;
			}
		}
```

and it gave this output
```
2025-04-13 07:08:46 DEBUG   192.168.127.1        77     33 integer and fk_source
2025-04-13 07:08:46 DEBUG   192.168.127.1        77     33 string and sourcetype
2025-04-13 07:08:46 DEBUG   192.168.127.1        77     33 integer and fk_target
2025-04-13 07:08:46 DEBUG   192.168.127.1        77     33 string and targettype
2025-04-13 07:08:46 DEBUG   192.168.127.1        77     33 string and relationtype
```
so it seems like the code works as intended?

Here is the request url
```
http://localhost:22080/api/index.php/objectlinks?fk_source=49&sourcetype=subscription&fk_target=227&targettype=facture
```

JonBendtsen replied:
@mdeweerd  I flipped the IF above such that string came first, hoping it would give a different PHPstan error, but it does not, same error :-/

Would you mind taking another look at provide some more guidance?

mdeweerd replied:
@JonBendtsen Did you see that I had already posted a PR.

I think that the array must be local and not through a constant so that phpstan uses the values.
The string[] type for the variable is certainly hurting, but converting to a constant and removing the typing hint did not help.
The working approach uses a private function to avoir repetition rather than using a static variable or constant.

JonBendtsen replied:
@mdeweerd  sorry, I had not noticed, I'll pay better attention in the future.

I like your approach, let's see if `Stan` does too :-)

JonBendtsen replied:
@mdeweerd  thanks for your help :-)

mdeweerd commented:
@JonBendtsen #33757 may interest you and maybe you have some suggestions.

eldy committed:
Merge branch 'develop' into from_develop_api_objectlinks

Commit:
bringing $FIELDS up to develop branch standard

Commit:
trying to tell phan the expected array types

Commit:
phan return and input type specifications

Commit:
trying to fix 'Method ObjectLink::_makeobject() should return int but return statement is missing.'

Commit:
trying to return an ObjecTLink

Commit:
get returns an ObjectLink object

Commit:
making sure that function _makeobject always has an return

Commit:
making sure deleteByValues always has a return statement

Commit:
remove empty line

Commit:
making sure getByValues always has a return value

Commit:
trying to fix phan for _validate

Commit:
no < around objectlink

Commit:
trying to fix some phan and phan-stan errors

Commit:
public variable target should be fk_target

Commit:
making TRIGGER_PREFIX more like in core/class/commonobject.class.php

Commit:
trying to save and use the notrigger value

Commit:
typecast objectlink fk_source and fk_target to string in the hope that they are later made correctly to int

JonBendtsen commented:
just trying to get it to build the latest commit, thus the "read for review"

Commit:
whitespace to trigger a build of my branch

JonBendtsen committed:
Merge branch 'develop' into from_develop_api_objectlinks

Commit:
back to being int - because that is what they are

Commit:
trying to typecast

Commit:
trying to help PHPStan to validate fk_source and fk_target

Commit:
trying to use  the same way as  are used

JonBendtsen committed:
Merge branch 'develop' into from_develop_api_objectlinks

JonBendtsen committed:
Merge branch 'develop' into from_develop_api_objectlinks

Commit:
splitting  over multiple lines

mdeweerd committed:
Fix: update ObjectLink return type in API methods

### FIX: Update ObjectLink return type in API methods

- Updated the return type of several methods from `ObjectLink` to `Object` to reflect a return type compatible with Luracast/Restler

JonBendtsen committed:
Merge pull request #2 from mdeweerd/pr33781

Fix: update ObjectLink return type in API methods

mdeweerd committed:
fix: refactor field setting in ObjectLinks class

# Fix: Refactor field setting in ObjectLinks class

- Removed the static `$INTFIELDS` array and moved its logic to a new private
method `_setObjectLinkField`.
- Updated `create`, `getByValues`, and `deleteByValues` methods to use
`_setObjectLinkField` for setting fields.
- Cleaned up the code by removing redundant field validation and type casting
logic.

Commit:
noting that relation type can both be string and null

JonBendtsen committed:
Merge branch 'develop' into from_develop_api_objectlinks

Commit:
escaping strings before parsing them to the database

Commit:
-> not global

JonBendtsen committed:
Merge branch 'develop' into from_develop_api_objectlinks

JonBendtsen committed:
Merge branch 'develop' into from_develop_api_objectlinks

Commit:
trying to fix phpstan errors by flipping so the string part is first

mdeweerd committed:
Merge branch 'fixme2' into pr33781

JonBendtsen committed:
Merge branch 'develop' into from_develop_api_objectlinks

JonBendtsen committed:
Merge pull request #3 from mdeweerd/pr33781

Fix: Refactor field cleaning in ObjectLinks class

JonBendtsen committed:
Merge branch 'develop' into from_develop_api_objectlinks

JonBendtsen committed:
Merge branch 'develop' into from_develop_api_objectlinks

JonBendtsen committed:
Merge branch 'develop' into from_develop_api_objectlinks

JonBendtsen committed:
Merge branch 'develop' into from_develop_api_objectlinks

JonBendtsen committed:
Merge branch 'develop' into from_develop_api_objectlinks

JonBendtsen commented:
@eldy  I believe this one is ready, it just build cleanly :-)
